cats/optimise_starttime.py

```diff
@@ -0,0 +1,91 @@
+from datetime import datetime, timezone
```

Comment:
I'll add my general comment on this module here so we can have a thread.

I think this PR adds useful missing pieces but I'm a bit uncomfortable with the fast it completely removes the `timeseries_conversion` and `forecast` modules. Makes more sense to me to build upon them instead of discarding them alltogether: they already contain most of the functionality (re)implemented in this module. Particularly, I'm not sure why `WindowedForecast` disappeared -- it was merged 2 days ago -- see #36 . I think it would be straightforward to make it work with a simple summation of intensity values.

That being a few welcome things in this module that are missing in the current implementation:

- Detecting cases where the jon duration exceeds the maximum forecast duration.
- Allowing to compute the windowed/summed/integrated intensity over the window by just summing intervals.
- Handle the fact that the first and last interval in the time window are fraction of the default interval length (30min).

Again -- I think we'd be better off integrating the above in the current code, rather than reimplementing a whole new module from scratch. Unless of course we think that there are strong reasons to do so, but not obvious to me at this stage.

Reply:
Regarding `WindowedForecast`: should definitely be kept, as highlighted in the PR release note and in the code [here](https://github.com/GreenScheduler/cats/blob/8c927c7c6090c1baa8fc018d802909900c6d9b23/cats/optimise_starttime.py#L38), the goal is to keep the trapezoidal integration as an option, especially while we figure out which one is right for the UK (issue #42). But it's a rather delicate function to get the bounds right (especially first and last segments), so rather than just shoehorn in into this updated structure, it'd be good to discuss about the best way to include it.

Regarding `forecast` it's not removed, just renamed as I'm not sure `forecast` still describes what this part of the code is doing. But it still serves the same purpose. 

As of `timeseries_conversion` (and most of the code in this PR), it's of course reusing existing code, sometimes grouped differently based on how usage has evolved. Which functions in particular from `timeseries_conversion` have been removed completely but still needed? My understanding is that `csv_loader` and `cat_converter` aren't needed anymore (as per PR #41), `check_duration` has just been moved to `check_clean_arguments` as it's what it does and `get_lowest_carbon_intensity` is now in `optimise_starttime` for the same reason.

Reply:
> the goal is to keep the trapezoidal integration as an option [..]. But it's a rather delicate function to get the bounds right (especially first and last segments), so rather than me just shoehorn in into this updated structure on my own, it'd be good to discuss about the best way to include it.

Don't you think it will more effort re-integrating something that is already there (and covered by a couple of tests) into new code? As opposed to start from what is there. Sorry -- I understand the appeal of writing that module from scratch, but I'm questioning whether it will make the next few steps actually harder. Obviously that's just my own humble opinion, let's see what other people have to say about this!

> Regarding forecast it's not removed, just renamed as I'm not sure forecast still describes what this part of the code is doing. But it still serves the same purpose.

Did `forecast` become `optimise_starttime`? It looks like a new module.

Reply:
I'll submit a commit later today that reintroduces `WindowedForecast` closer to its original form so that we can hopefully get the best of both worlds

Reply:
Turns out that properly integrating the first and last time windows into the trapezoidal model was a bit harder than planned (didn't manage with the original implementation but this one is very close) but it's now done with a test: …
[1aca7ab](https://github.com/GreenScheduler/cats/pull/43/commits/1aca7abb0331de809bd9b83f74b48a5fc18a0aa9)

Reply:
Also @tlestang if you think testing the integration method on the sinus function is still needed, it can be put back in as it's only commented out for now, for now there is a test on a dummy carbon intensity forecast